src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.start).before(dist)
+
+    def after(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.stop - 1).after(dist)
+
+    def prev(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes leading up to the start?
+        raise NotImplementedError('Block.prev')
+
+    def next(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes past the end?
+        raise NotImplementedError('Block.next')
+
+    # ------------------------------------------------------------------------ #
+    # Sequence interface implementation
+    # ------------------------------------------------------------------------ #
+
+    def __iter__(self):
+        attr, _range = self._path[-1]
+        block = self.parent()
+        for i in _range:
+            yield block.child(attr, i)
+
+    def __getitem__(self, i):
+        attr, r = self._path[-1]
+        r = r[i]
+        if isinstance(r, range):
+            if r.step != 1:
+                raise IndexError('block cursors must be contiguous')
+            return Block(self._proc, self._path[:-1] + [(attr, r)])
+        else:
+            return self.parent().child(attr, r)
+
+    def __len__(self):
+        _, _range = self._path[-1]
+        return len(_range)
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _replace(self, nodes: list, *, empty_default=None):
+        """
+        This is an UNSAFE internal function for replacing a block in an AST with
+        a list of statements and providing a forwarding function as collateral.
+        It is meant to be package-private, not class-private, so it may be
+        called from other internal classes and modules, but not from end-user
+        code.
+        """
+        assert self._path
+        assert len(self) > 0
+
+        def update(node):
+            attr, i = self._path[-1]
+            children = getattr(node, attr)
+            new_children = children[:i.start] + nodes + children[i.stop:]
+            new_children = new_children or empty_default or []
+            return node.update(**{attr: new_children})
+
+        from .API import Procedure
```

Comment:
Would it be better to `from . import API` at the top of this file and then just call `API.Procedure` below?

Reply:
btw, if you've never read Craig Chambers' papers on efficient implementation of the language `Self`, you might find those interesting w.r.t. these kinds of issues in dynamic language implementation

Reply:
IIRC it was causing a circular import error, but that might have been fixed now that `API` doesn't import `cursors` anymore.

Reply:
Maybe it was `from .API import Procedure` at the top level?  That can often cause circular import errors.  As a first recourse, I recommend switching to `from . import API` since simply importing the module will much more rarely cause the circular import problems.

Reply:
Nope, I still get an error here:
```
tests/conftest.py:19: in <module>
    from exo import Procedure, compile_procs
src/exo/__init__.py:2: in <module>
    from .API import Procedure, compile_procs, proc, instr, config
src/exo/API.py:19: in <module>
    from .pattern_match import match_pattern, get_match_no, match_cursors
src/exo/pattern_match.py:14: in <module>
    from .cursors import Cursor, Node, Block
src/exo/cursors.py:11: in <module>
    from .API import Procedure
E   ImportError: cannot import name 'Procedure' from partially initialized module 'exo.API' (most likely due to a circular import) (/home/alex/Development/exo/src/exo/API.py)
```

Reply:
Yes, you're trying to do `from .API import Procedure`.  I'm saying, yes don't do that; it's not safe circularly.  Instead do
`from . import API`
at the top level, rather than putting an import statement in the middle of a method.

Reply:
Doh, that fixes it, thanks